data/bookkeeping/block.go

```diff
@@ -364,18 +382,21 @@ func ProcessUpgradeParams(prev BlockHeader) (uv UpgradeVote, us UpgradeState, er
 	// If there is no upgrade proposal, see if we can make one
 	if prev.NextProtocol == "" {
 		for k, v := range prevParams.ApprovedUpgrades {
-			if v {
-				upgradeVote.UpgradePropose = k
-				upgradeVote.UpgradeApprove = true
-				break
-			}
+			upgradeVote.UpgradePropose = k
+			upgradeVote.UpgradeDelay = basics.Round(v)
+			upgradeVote.UpgradeApprove = true
+			break
 		}
 	}
 
-	// If there is a proposal being voted on, see if we approve it
+	// If there is a proposal being voted on, see if we approve it and its delay
 	round := prev.Round + 1
 	if round < prev.NextProtocolVoteBefore {
-		if prevParams.ApprovedUpgrades[prev.NextProtocol] {
+		expectedDelay, ok := prevParams.ApprovedUpgrades[prev.NextProtocol]
+		if expectedDelay == 0 {
+			expectedDelay = prevParams.DefaultUpgradeWaitRounds
+		}
+		if ok && prev.NextProtocolVoteBefore+basics.Round(expectedDelay) == prev.NextProtocolSwitchOn {
```

Comment:
That's somewhat weird behavior: a node won't vote yes unless the proposer picked the same UpgradeWaitRounds as this node?  What's the rationale?

Reply:
I was thinking it would be good to avoid the scenario where a proposer tries to schedule a more urgent upgrade with a large delay or a slow upgrade with a short delay. I don't know how much this matters.

Reply:
Either way seems potentially reasonable, but if there's no flexibility in terms of the number of wait rounds, then it seems like this should just be a fixed value in the next protocol's `ConsensusParams`, and then we don't need any min/max values.